walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -47,16 +47,36 @@
     }
 
     // Add optional guests
-    // TODO: Use stream to make a deep copy
-    List<TimeRange> optionalMeetingTimes = new ArrayList<TimeRange>();
-    for (int i = 0; i < meetingTimes.size(); i++) {
-      optionalMeetingTimes.add(meetingTimes.get(i).fromStartDuration(meetingTimes.get(i).start(), 
-          meetingTimes.get(i).duration()));
-    }
-    optionalMeetingTimes = findSchedule(possibleTimes, events, optionalGuests, request);
-    if (!optionalMeetingTimes.isEmpty() || attendees.isEmpty()) {
+    List<TimeRange> optionalMeetingTimes = copyListTimeRnage(meetingTimes);
+    optionalMeetingTimes = findSchedule(timeRangeToRangeSet(meetingTimes), events, optionalGuests, request);
+    if (!optionalMeetingTimes.isEmpty()) {
       return optionalMeetingTimes;
     }
+
+    // find most optional guests that can attend
+    if (optionalMeetingTimes.isEmpty()) {
```

Comment:
If you want to take this to the EXSTREAM, think about how you could make this _entire method_ an efficient ~one-liner (barring some input validation and initial setup) :)

I think a one-liner version could definitely get into ugly code territory, and am not sure if I'd recommend keeping a stream-based solution, but I think it would be a fun exercise / challenge for you if you're up for it. The key to making it readable is how you break out your helper methods.

A couple of hints:

1. Your stream will be what you're already iterating over, subsets of optional guests.
2. You only care about the subset with the maximal amount. `Stream.first` returns an `Optional` containing the first element of a stream. Is there a way to use this to efficiently pick out the best set of meeting times? (other relevant functions: `Stream.map`, `Stream.filter`)
3. Your method currently breaks out several different cases explicitly, can you think of a way to handle this a bit more naturally / implicitly? Note: this will likely become apparent once you shift your perspective to streams using the first two hints.
    1. There are no mandatory or optional attendees.
    2. There are mandatory but no optional attendees.
    3. There are valid meeting times that include all optional attendees.
    4. There is at least one subset of optional attendees such that a valid schedule exists (note that your current strategy includes the _empty set_ of optional attendees).

Reply:
Here's my intended solution:

```java
public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
  // Input sanitization / edge case, make sure the request is less than a day.
  if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
    return Arrays.asList();
  }

  // 1. Get set of possible times with respect to mandatory attendees
  RangeSet<Integer> allDay = TreeRangeSet.create();
  allDay.add(Range.closedOpen(0, 24 * 60));
  // Change the signature of findSchedule to return a RangeSet.
  RangeSet<Integer> possibleTimes = findSchedule(allDay, events, request.getAttendees(), request);

  // 2. Get a stream of all subsets sorted by descending cardinality.
  return allSubsetsDescending(request.getOptionalAttendees())
    // 3. Find the possible times for a given subset.
    .map(optionalAttendees -> findSchedule(possibleTimes, events, optionalAttendees, request))
    // 4. Filter out empty schedules due to impossible subsets.
    .filter(Predicates.not(RangeSet::isEmpty))
    // 5. Convert back to the desired output format.
    .map(FindMeetingQuery::rangeToTimeRange)
    // 6. Take the first element of the stream. Note: since we're streaming by descending cardinality,
    // the first element corresponds to the highest cardinality subset that had possible times.
    .first()
    // 7. Retrieve the first element, or return the empty list if the stream was empty.
    .orElse(Arrays.asList());
}
```

This handles all of the different cases I mentioned above implicitly:

1. There are no mandatory or optional attendees.
    * `possibleTimes` will be all day, and the empty set of optional attendees will be the only element of the stream.
2. There are mandatory but no optional attendees.
    * The empty set will be the only element of the stream, yielding `possibleTimes`.
3. There are valid meeting times that include all optional attendees.
    * The descending iteration order will cause the pipeline to only look at that single full set and return it immediately.
4. There is at least one subset of optional attendees such that a valid schedule exists.
    * The empty set is included in "all subsets", meaning `possibleTimes` is will be returned, as intended.

This approach has the same time complexity but since we only stream the subsets it saves space on the order of 2^(number of optional attendees).

`allSubsetsDescending` will be a little tricky to implement on your own but an alternative is to use `Stream.max` along with annotating subsets with their size with a custom datatype.

Guava provides `Sets.powerSet`, but to get the descending behavior we'd probably want to use `Sets.combinations`.

In production code I would probably go with the `Stream` approach and use `Sets.combinations` in a helper.